keysign/send.py

```diff
@@ -84,72 +83,155 @@ def __init__(self, builder=None):
+            info = yield self.offer.allocate_code(worm=False)
+            code, discovery_data = info
+            self.create_keypresent(code, discovery_data)
+            defers = self.offer.start()
+            for de in defers:
+                de.addCallback(self._received)
+
+    def _received(self, start_data):
+        success, message = start_data
+        if message and type(message) == LonelyError:
+            # This only means that we closed wormhole before a transfer
+            pass
+        elif message and message == "Back":
+            # Simply the return of the back button, no errors here
+            pass
+        elif message and type(message) == ServerConnectionError:
+            self._deactivate_timer()
+            self.deactivate()
+            self.klw.code_spinner.stop()
+            self.no_connection()
         else:
-            log.info("Bluetooth as been skipped")
+            self.show_result(success, message)
+
+    def slow_connection(self):
+        self.klw.label_ib.set_label("Very slow Internet connection!")
+        self.klw.ib.show()
+        log.info("Slow Internet connection")
+
+    def no_connection(self):
+        self.klw.label_ib.set_label("No Internet connection!")
+        self.klw.ib.show()
+        log.info("No Internet connection")
+
+    def create_keypresent(self, code, discovery_data):
+        self._deactivate_timer()
         log.info("Use this for discovering the other key: %r", discovery_data)
         ####
-        # Create and show widget for key
-        kpw = KeyPresentWidget(key, qrcodedata=discovery_data)
-        self.stack.add(kpw)
+        # Create widget for key
+        self.kpw = KeyPresentWidget(self.key, code, discovery_data)
+
+        ####
+        # Show widget for key
+        self.stack.add(self.kpw)
         self.stack_saved_visible_child = self.stack.get_visible_child()
-        self.stack.set_visible_child(kpw)
-        log.debug('Setting kpw: %r', kpw)
-        self.kpw = kpw
+        self.stack.set_visible_child(self.kpw)
+        log.debug('Setting kpw: %r', self.kpw)
+        self.klw.ib.hide()
+        self.klw.code_spinner.stop()
 
-    def deactivate(self):
-        self.stopped = True
+    def show_result(self, success, message):
         self._deactivate_offer()
 
-        ####
-        # Re-set stack to inital position
-        self.stack.set_visible_child(self.stack_saved_visible_child)
+        self.stack.add(self.rb)
         self.stack.remove(self.kpw)
         self.kpw = None
-        self.stack_saved_visible_child = None
+
+        if success:
+            self.result_label.set_label("Key successfully sent.\n"
+                                        "You should receive soon an email with the signature.")
+            self.stack.set_visible_child(self.rb)
+        else:
+            if type(message) == WrongPasswordError:
+                self.result_label.set_label("The security of the connection seems low.\n"
```

Comment:
user facing, no?

Also.. maybe.. "Could not establish a secure connection. ..." instead of "low" security.

Reply:
Done with e5bfb2d